### PR TITLE
Add alerting & recording rules, series deletion & retention docs

### PR DESCRIPTION
**Pull request by VineethReddy02**
This PR contains docs for

1. Setting up alerting, recording rules at the Prometheus side. 
2. Metric deletion & metric retention using Promscale & TimescaleDB.

### CODE REVIEW

File: docs/alerting-recording.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# Alerting & Recording Rules
+
+### Alerting Rules
+
+The alerting rules are used to trigger alerts based on the provided alerting rules an alert should be fired to an external service like slack. In promscale the alerting rules are evaluated in prometheus. i.e define alerting rules in prometheus config
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:5]: 1. This sentence does not sound good.
2. We do not define alerting rules in the prometheus configuration file. We mention the alerting-rules file name as a list where these files contains the actual rules. This statement needs an update in this direction.

Comment (Harkishen-Singh) [reply docs/alerting-recording.md:5]:
```suggestion
Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done on the Prometheus side.
```

File: docs/alerting-recording.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      expr: <promql_expression>
+      for: <time-interval for how long this to happen to happen to fire an alert>
+      labels:
+        <key>:<value>
+      annotations:
+        summary: <text>
+        description: <description on alert>
+```
+
+Load the above defined alert rule file to your prometheus config for evaluation. i.e
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:21]: Maybe, have this statement linked with the above one after updating.

File: docs/alerting-recording.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  alertmanagers:
+  - static_configs:
+    - targets:
+      - localhost:9093
+```
+
+
+### Recording Rules
+
+The recording rules are for pre-calculating frequently used or computationally expensive queries. The results of those rules are saved into their own time series. In promscale the rules rules are evaluated in prometheus. i.e define recording rules in prometheus config
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:41]: please mention that these newly formed time-series follow the name of the record (please confirm once)

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]

File: docs/alerting-recording.md (hunk 2 above)
Comment (VineethReddy02) [reply docs/alerting-recording.md:21]: The sequence is to add alerting rules in a file, define a reference to the rules file, and configure the alert manager endpoint.

Comment (Harkishen-Singh) [reply docs/alerting-recording.md:21]: Can we please change `config` -> `configuration` and `i.e` -> `i.e.,` (along with full-stops where needed) across the entire file?

File: docs/alerting-recording.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
+# Alerting & Recording Rules
+
+### Alerting Rules
+
+Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done on the Prometheus side.
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:5]:
```suggestion
Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the alert manager (link). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done at the Prometheus side.
```

1. Does this sound better with `at`?
2. Please add the link to the alert manager.

File: docs/metric_deletion_retention.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+SQL cmd to set the retention policy for a specific metric
+
+```
+SELECT prom_api.set_metric_retention_period(container_cpu_usage_seconds_total, INTERVAL '1 day' * 2)
+```
+
+SQL cmd to reset specific metric retention to the default metric retention
+
+
```
Comment (Harkishen-Singh) [inline docs/metric_deletion_retention.md:42]: I think this line got added in a mistake. (trailing line)

File: docs/metric_deletion_retention.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+```
+SELECT prom_api.reset_metric_retention_period(container_cpu_usage_seconds_total)
+```
+
+The retention policies are executed using a cron job. The details for setting up a crontab can be found [here](https://github.com/timescale/promscale/blob/master/docs/binary.md#-setting-up-cron-jobs)
```
Comment (Harkishen-Singh) [inline docs/metric_deletion_retention.md:47]:
```suggestion
The retention policies are executed using a cron job. The details for setting up a crontab can be found [here](https://github.com/timescale/promscale/blob/master/docs/binary.md#-setting-up-cron-jobs).
```

File: docs/alerting-recording.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+Load the above defined recording rule file to your prometheus config for evaluation. i.e 
+
+```
+rule_files:
+  - "<recording-rules-file>"
+```
+
+More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/)
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:63]:
```suggestion
More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).
```

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
Sorry, just last set of nits.

File: docs/metric_deletion_retention.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+Promscale exposes an API for deletion of series. This works same as prometheus delete API.
+
+### Promscale Delete API
+
+URL query parameters:
+
+* **match[]=<series_selector>**: Repeated label matcher argument that selects the series to delete. At least one match[] argument must be provided.
+* **start=<rfc3339 | unix_timestamp>**: Start timestamp. Optional and defaults to minimum possible time.
+* **end=<rfc3339 | unix_timestamp>**: End timestamp. Optional and defaults to maximum possible time.
```
Comment (cevian) [inline docs/metric_deletion_retention.md:11]: @Harkishen-Singh the start and end options are not supported yet, correct? We should take them out of here then and put a note at the top of this section.

File: docs/alerting-recording.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    - targets:
+      - localhost:9093
+```
+
+More details on alerting rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/).
+
+
+### Recording Rules
+
+The recording rules are for pre-calculating frequently used or computationally expensive queries. The results of those rules are saved into their own time series. The newly formed time-series will follow the name of the record. In promscale the rules are evaluated in prometheus i.e define recording rules in prometheus configuration.
```
Comment (cevian) [inline docs/alerting-recording.md:43]: these rules are fired in prometheus but will they fetch data from promscale if needed or does the data retention in prometheus have to be long enough to contain all the data. we should clarify.

Also we should clarify that the new series is written into promscale.

File: docs/metric_deletion_retention.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+PUT /delete_series
+```
+
+Example:
+
+```
+curl -X POST -g http://promscale:9201/delete_series?match[]=container_cpu_usage_seconds_total
+```
+
+# Metric Retention
```
Comment (cevian) [inline docs/metric_deletion_retention.md:24]: We should add a todo to support metric (not series) deletion

File: docs/metric_deletion_retention.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
+# Metric Deletion
```
Comment (cevian) [inline docs/metric_deletion_retention.md:1]:
```suggestion
# Series Deletion
```

Comment (cevian) [inline docs/metric_deletion_retention.md:1]: I'd add a paragraph above as an intro outlining the different types of deletions we support. Namely deletion of entire series or deletion of old data in individual metrics.

File: docs/alerting-recording.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
+# Alerting & Recording Rules
+
+### Alerting Rules
+
+Alerting rules are used to trigger alerts based on the violation of any condition(s). These alerts are fired to external services like slack, mails, etc by the [alert manager](https://prometheus.io/docs/alerting/latest/alertmanager/). Alerting rules are written in a YAML file and paths of these files are mentioned in the Prometheus configuration respectively. It is important to note that the evaluation of these conditional rules are done at the Prometheus side.
```
Comment (cevian) [inline docs/alerting-recording.md:5]: We should clarify whether this will fetch data from promscale if needed or not

**Review by cevian — CHANGES_REQUESTED** [review]

**Comment by cevian** [comment]
We should also add links to these pages from the front-page Table of contents in this PR.

File: docs/metric_deletion_retention.md (hunk 8 above)
Comment (Harkishen-Singh) [reply docs/metric_deletion_retention.md:11]: Yes, they are not supported. We delete the complete series that match the matchers. (Provided in the `match[]` param)

File: docs/metric_deletion_retention.md (hunk 10 above)
Comment (VineethReddy02) [reply docs/metric_deletion_retention.md:24]: Currently, deletion API supports metric & series deletion, right? 
IMO 
metric: `go_goroutines`
series: `go_goroutines{instance: "localhost:9090"}`

https://stackoverflow.com/questions/49859360/how-do-i-delete-a-time-series-from-prometheus-v2-specifically-a-series-of-alert

File: docs/alerting-recording.md (hunk 9 above)
Comment (VineethReddy02) [reply docs/alerting-recording.md:43]: If `read_recent: true` is configured under `remote-read` configuration of prometheus. Every read request is forwarded to remote-read. The alerting & recording rules evaluation is performed by querying the data from remote-read. 

If disabled the data from promethues will be used. In both alerting & recording rules the evaluation happens for every `15s` or can be configured and the new series are written to promscale. The data retention policies configured should be higher than the time interval specified in alerting & recording rules (only if `read_recent` is disabled.)

Comment (Harkishen-Singh) [reply docs/alerting-recording.md:43]: Then let's mention users use `read_recent: true` as recommended over here and in the `note` section.

File: docs/alerting-recording.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+Load the above defined recording rule file to your prometheus configuration for evaluation.
+
+```
+rule_files:
+  - "<recording-rules-file>"
+```
+
+More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).
+
+**Note**: If `read_recent: true` is set in prometheus configuration the data is fetched from promscale for evaluation. If `read_recent` is disabled the data stored in prometheus will be used for evaluation this needs prometheus retention to be configured higher than the alerting & recording rules evaluation interval for the rules evaluation to happen before dropping the samples. The new alerting & recording rules series are written to promscale.
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:65]: Let's mention here that `read_recent: true` is the recommended way.

Comment (Harkishen-Singh) [reply docs/alerting-recording.md:65]: This statement can be modified to sound better. Also, the second sentence is too long.

File: docs/metric_deletion_retention.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
+Promscale supports series deletion & TimescaleDB supports metric retention. The series deletion is performed on the complete series, whereas the metric retention can be enforced on all the metrics or per metric basis.
```
Comment (Harkishen-Singh) [inline docs/metric_deletion_retention.md:1]: **TimescaleDB supports**

I don't think this is required. WDYT @cevian?

Comment (Harkishen-Singh) [reply docs/metric_deletion_retention.md:1]:
```suggestion
Promscale supports series deletion & metric retention. The series deletion is performed on the entire time-range, whereas the metric retention can be enforced on all the metrics or per metric basis.
```

File: docs/metric_deletion_retention.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+## Series Deletion
+
+Promscale exposes an API for deletion of series. This works same as prometheus delete API.
+
+**Note**: The start and end timestamp options are not supported yet. Currently, the `delete_series` API lets you to delete complete series.
```
Comment (Harkishen-Singh) [inline docs/metric_deletion_retention.md:7]: I do not understand the meaning of **complete series** (opposite, partial series?). It should be **lets you delete the series across the entire time-range.**

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
I would not advise using (any) short forms in docs. One that I noticed is `cmd`. Let's make that `command`.

File: docs/metric_deletion_retention.md (hunk 10 above)
Comment (Harkishen-Singh) [reply docs/metric_deletion_retention.md]: > We should add a todo to support metric (not series) deletion

@cevian Why is that required? Can't just users do `match[]={__name__="<metric_name>"}` or simply `match[]=<metric_name>`?

File: docs/alerting-recording.md [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    - targets:
+      - localhost:9093
+```
+
+More details on alerting rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/).
+
+
+### Recording Rules
+
+The recording rules are for pre-calculating frequently used or computationally expensive queries. Recording rules are evaluated at the Prometheus side i.e., defined in recording rules in the Prometheus configuration. The evaluated rules form a new series which go by the name of the record field's value and ingested in both Prometheus and Promscale.
```
Comment (Harkishen-Singh) [inline docs/alerting-recording.md:43]:
```suggestion
The recording rules are for pre-calculating frequently used or computationally expensive queries. PromQL expressions defined in recording rules are evaluated at the Prometheus side. These evaluated rules form a new series that goes by the name of the record field's value and ingested in both Prometheus and Promscale.
```

**Review by Harkishen-Singh — APPROVED** [review]
Lgtm :+1:. You can merge this after Mat's approval.

File: docs/alerting-recording.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+Load the above defined recording rule file to your prometheus configuration for evaluation.
+
+```
+rule_files:
+  - "<recording-rules-file>"
+```
+
+More details on recording rules can be found [here](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).
+
+**Note**: The recommended way is to enable `read_recent` to `true` in the Prometheus configuration. This lets you fetch the data from Promscale for evaluation of PromQL queries. If `read_recent` is disabled, the data stored in Prometheus's local tsdb will be used for evaluating alerting/recording rules. Evaluation of rules on Prometheus's local tsdb will have dependency on Prometheus's retention time and not by Promscale.
```
Comment (cevian) [inline docs/alerting-recording.md:65]:
```suggestion
**Note**: We recommended setting `read_recent` to `true` in the [Prometheus remote_read configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read) when using recording rules. This tells Prometheus to fetch data from Promscale when evaluating PromQL queries (including recording rules). If `read_recent` is disabled, **only** the data stored in Prometheus's local tsdb will be used when evaluating alerting/recording rules and thus be dependent on the retention period of Prometheus (not Promscale).
```

File: docs/metric_deletion_retention.md [hunk 18]
```diff
@@ -0,0 +1,49 @@
+Promscale supports series deletion & metric retention. The series deletion is performed on the entire time-range, whereas the metric retention can be enforced on all the metrics or per metric basis.
```
Comment (cevian) [inline docs/metric_deletion_retention.md:1]:
```suggestion
# Deleting Data

Promscale supports series deletion & metric retention. Series deletion deletes all data for a specified series over the entire time-range. Metric retention deletes data older than a specified duration for a given metric. You can specify a default metric retention period as well as overwrite that default on a particular metric.  
```

File: docs/metric_deletion_retention.md [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+Example:
+
+```
+curl -X POST -g http://promscale:9201/delete_series?match[]=container_cpu_usage_seconds_total
+```
+
+## Metric Retention
+
+TimescaleDB offers full control over data retentions i.e data retention policies can be applied on the complete series ingested and also per metric basis using SQL queries.
```
Comment (cevian) [inline docs/metric_deletion_retention.md:28]:
```suggestion
TimescaleDB offers full control over data retentions i.e you can set a default data retention period as well as overwrite the default on a per-metric basis.
```

File: docs/metric_deletion_retention.md [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+SQL command to set the default retention policy to two days for all series.
```
Comment (cevian) [inline docs/metric_deletion_retention.md:30]:
```suggestion
SQL command to set the default retention policy to two days for all metrics that do not have custom retention policy already specified.
```

File: docs/metric_deletion_retention.md [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+```
+SELECT prom_api.set_default_retention_period(INTERVAL '1 day' * 2)
+```
+
+SQL command to set the retention policy for a specific metric.
```
Comment (cevian) [inline docs/metric_deletion_retention.md:36]:
```suggestion
SQL command to set a custom retention policy for a specific metric.
```

File: docs/metric_deletion_retention.md [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+SELECT prom_api.set_metric_retention_period(container_cpu_usage_seconds_total, INTERVAL '1 day' * 2)
+```
+
+SQL command to reset specific metric retention to the default metric retention.
+
+```
+SELECT prom_api.reset_metric_retention_period(container_cpu_usage_seconds_total)
+```
+
+The retention policies are executed using a cron job. The details for setting up a crontab can be found [here](https://github.com/timescale/promscale/blob/master/docs/binary.md#-setting-up-cron-jobs).
```
Comment (cevian) [inline docs/metric_deletion_retention.md:48]:
```suggestion
The retention policies are executed using a cron job. Please see the Promscale installation instructions for you platform to see how to set up the cron job. 
```

File: README.md [hunk 23]
```diff
@@ -55,6 +55,8 @@ remote storage endpoints. For more information about prom-migrator, visit
   * [Filtering Series](docs/sql_schema.md#filtering-series)
   * [Data Retention](docs/sql_schema.md#data-retention)
 * **[Writing Data to Promscale](docs/writing_to_promscale.md)**
+* **[Alerting & Recording Rules](docs/alerting-recording.md)**
+* **[Series Deletion & Metric Retention](docs/metric_deletion_retention.md)**
```
Comment (cevian) [inline README.md:59]:
```suggestion
* **[Deleting Data](docs/metric_deletion_retention.md)**
```

**Review by cevian — APPROVED** [review]
Have some comments. If you are ok with my suggestions feel free to merge (no need for re-review). If you want to change anything please ping me for another review

**Pull request merged**